pkg/frontend/querymiddleware/roundtrip.go

```diff
@@ -176,8 +177,9 @@ func NewTripperware(
 	cacheExtractor Extractor,
 	engineOpts promql.EngineOpts,
 	registerer prometheus.Registerer,
+	frontendState func() services.State,
```

Comment:
wdyt about using an interface here instead of a function signature? Interface implementation are easier to track with an IDE and functions require you to dig through all the callsites.

Reply:
Makes sense to me, I've added this in 7df361022cb86189fbadcf0935452184fbc572dc.